PEP 669: Refinements and clarifications (#3021)
python/peps: pep-0669.rst

* PY_RETURN: Return from a Python function (occurs immediately before the return, the callee's frame will be on the stack).
* PY_YIELD: Yield from a Python function (occurs immediately before the yield, the callee's frame will be on the stack).
* PY_UNWIND:  Exit from a Python function during exception unwinding.
* CALL: A call in Python code (event occurs before the call).
* C_RETURN: Return from any callable, except Python functions (event occurs after the return).
* C_RAISE: Exception raised from any callable, except Python functions (event occurs after the exit).
* RAISE: An exception is raised, except those that cause a ``STOP_ITERATION`` event.
* EXCEPTION_HANDLED: An exception is handled.
* LINE: An instruction is about to be executed that has a different line number from the preceding instruction.
* INSTRUCTION -- A VM instruction is about to be executed.
* JUMP -- An unconditional jump in the control flow graph is made.
* BRANCH -- A conditional branch is taken (or not).

> * STOP_ITERATION -- An artificial ``StopIteration`` is raised, see below.

Reviewer: ```suggestion
* STOP_ITERATION -- An artificial ``StopIteration`` is raised; see :ref:`669-stopiteration`.
```

Name and link the relevant section (and fix comma splice grammar error)

More events may be added in the future.

All events will be attributes of the ``events`` namespace in ``sys.monitoring``.
All events will represented by a power of two integer, so that they can be combined
with the ``|`` operator.

Events are divided into three groups:

Local events
''''''''''''

Local events are associated with normal execution of the program and happen
at clearly defined locations. All local events can be disabled.
The local events are:

* PY_START
* PY_RESUME
* PY_RETURN
* PY_YIELD
* CALL
* LINE
* INSTRUCTION
* JUMP
* BRANCH
* STOP_ITERATION

Ancilliary events
'''''''''''''''''

Ancillary events can be monitored like other events, but are controlled

> by another event.

Reviewer: ```suggestion
by another event:
```

* C_RAISE
* C_RETURN

The ``C_RETURN`` and ``C_RAISE`` events are are controlled by the ``CALL``
event. ``C_RETURN`` and ``C_RAISE`` events will only be seen if the
corresponding ``CALL`` event is being monitored.

Other events
''''''''''''

Other events are not necessarily tied to a specific location in the
program and cannot be individually disabled.

The other events that can be monitored are:

* PY_THROW
* PY_UNWIND
* RAISE
* EXCEPTION_HANDLED

> The STOP_ITERATION event

Reviewer: ```suggestion
.. _669-stopiteration:

The STOP_ITERATION event
```

Add cross-reference target label per above

When an active event occurs, the registered callback function is called.
Different events will provide the callback function with different arguments, as follows:

> * ``PY_START`` and ``PY_RESUME``:
> 
>     ``func(code: CodeType, instruction_offset: int) -> DISABLE | Any``

Reviewer: NB, all of these these would likely display better as code blocks (proper breaks, syntax highlighting, background, more easily copyable, no quote formatting/italics, etc) and with less formatting too. For example:

```suggestion
* ``PY_START`` and ``PY_RESUME``::

    func(code: CodeType, instruction_offset: int) -> DISABLE | Any
```

* ``PY_RETURN`` and ``PY_YIELD``:

    ``func(code: CodeType, instruction_offset: int, retval: object) -> DISABLE | Any``

* ``CALL``, ``C_RAISE`` and ``C_RETURN``:

>     ``func(code: CodeType, instruction_offset: int, callable: object, arg0: object | MISSING) -> DISABLE | Any``
>     If there are no arguments, ``arg0`` is set to ``MISSING``.

Reviewer: ```suggestion
    ``func(code: CodeType, instruction_offset: int, callable: object, arg0: object | MISSING) -> DISABLE | Any``

    If there are no arguments, ``arg0`` is set to ``MISSING``.
```

You probably want a line break here, so it doesn't get mashed together with the previous

tools. This could be useful for debugging a profiling tool, but would produce
misleading profiles, as the debugger tool would show up in the profile.

Order of events
---------------

If an instructions triggers several events they occur in the following order:

* INSTRUCTION
* LINE
* All other events (only one of these events can occur per instruction)

Each event is delivered to tools in ascending order of ID.

The "call" event group
----------------------

Most events are independent; setting or disabling one event has no effect on the others.
However, the ``CALL``, ``C_RAISE`` and ``C_RETURN`` events form a group.
If any of those events are set or disabled, then all events in the group are.

> Disabling a ``CALL`` event, will not disable the matching ``C_RAISE`` or ``C_RETURN``,

Reviewer: ```suggestion
Disabling a ``CALL`` event will not disable the matching ``C_RAISE`` or ``C_RETURN``,
```

Elide stray comma

In general these operations can be considered to be fast:

* ``def get_events(tool_id: int)->int``
* ``def get_local_events(tool_id: int, code: CodeType)->int``
* ``def register_callback(tool_id: int, event: int, func: Callable)->Optional[Callable]``
* ``def get_tool(tool_id) -> str | None``

These operations are slower, but not especially so:

* ``def set_local_events(tool_id: int, code: CodeType, event_set: int)->None``

> * ``def insert_marker(tool_id: int, code: CodeType, offset: int)->None``
> * ``def remove_marker(tool_id: int, code: CodeType, offset: int)->None``

Reviewer 2: There are no markers any more, right?
```suggestion
```

Author: Thanks, I thought I'd removed those.
